addons/binding/org.openhab.binding.zwave/src/main/java/org/openhab/binding/zwave/internal/protocol/commandclass/ZWaveClockCommandClass.java

```diff
@@ -64,10 +134,70 @@ public void handleApplicationCommandRequest(SerialMessage serialMessage, int off
         logger.debug("NODE {}: Received clock command (v{})", this.getNode().getNodeId(), this.getVersion());
         int command = serialMessage.getMessagePayloadByte(offset);
         switch (command) {
+            case CLOCK_REPORT:
+                int day = serialMessage.getMessagePayloadByte(offset + 1) >> 5;
+                int hour = serialMessage.getMessagePayloadByte(offset + 1) & 0x1f;
+                int minute = serialMessage.getMessagePayloadByte(offset + 2);
+                String days[] = new DateFormatSymbols().getWeekdays();
+                int javaDay = day == 7 ? 1 : day + 1;
+                logger.debug(String.format("NODE %d: Received clock report: %s %02d:%02d", getNode().getNodeId(),
+                        days[javaDay], hour, minute));
+
+                Calendar cal = Calendar.getInstance();
+                cal.set(Calendar.DAY_OF_WEEK, javaDay);
+                cal.set(Calendar.HOUR_OF_DAY, hour);
+                cal.set(Calendar.MINUTE, minute);
+                Date nodeTime = cal.getTime();
+                long clockOffset = (nodeTime.getTime() - System.currentTimeMillis()) / 1000 / 60;
+                ZWaveClockValueEvent zEvent = new ZWaveClockValueEvent(getNode().getNodeId(), endpoint, clockOffset);
+                this.getController().notifyEventListeners(zEvent);
+
+                Date low = new Date(System.currentTimeMillis() - CLOCK_PRESISION);
+                Date high = new Date(System.currentTimeMillis() + CLOCK_PRESISION);
+
+                // auto correct the time if off by more than 5 minutes
+                if (nodeTime.before(low) || nodeTime.after(high)) {
+                    updateClock = true;
+                }
+                break;
             default:
                 logger.warn(String.format("NODE %d: Unsupported Command %d for command class %s (0x%02X).",
                         this.getNode().getNodeId(), command, this.getCommandClass().getLabel(),
                         this.getCommandClass().getKey()));
         }
     }
+
+    @Override
+    public Collection<SerialMessage> getDynamicValues(boolean refresh) {
+        ArrayList<SerialMessage> result = new ArrayList<SerialMessage>();
+        if (refresh == true && getEndpoint() == null) {
+            result.add(getValueMessage());
+        }
+        if (updateClock) {
```

Comment:
I think I'd again be tempted to remove the updating of the time and do it in the application. It's more flexible if it's done there.